File: link-list/linked-list-values/linked-list-values.js
// Write a function, linkedListValues, that takes in the head of a linked list as an argument. The function should return an array containing all values of the nodes in the linked list.

class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

const fillValues = (head, values) => {
    if (head === null) return;
    values.push(head.val);
    fillValues(head.next, values);
};

const linkedListValues = head => {
    const newArr = [];
    fillValues(head, newArr);
    return newArr;
};

const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");

a.next = b;
b.next = c;
c.next = d;

// a -> b -> c -> d

console.log(linkedListValues(a)); // -> [ 'a', 'b', 'c', 'd' ]

File: 2.link-list/2.linked-list-values/linked-list-values.go
// Write a function, linkedListValues, that takes in the head of a linked list as an argument. The function should return an array containing all values of the nodes in the linked list.

package main

import "fmt"

type Node struct {
	value string
	next  *Node
}

func NewNode(value string) *Node {
	return &Node{value: value}
}

func fillValues(head *Node, arr *[]string) {
	if head == nil {
		return
	}
	*arr = append(*arr, head.value)
	fillValues(head.next, arr)
}

func linkedListValues(head *Node) []string {
	newArr := []string{}
	fillValues(head, &newArr) // Pass the address of the slice
	return newArr
}

func main() {
	a := NewNode("a")
	b := NewNode("b")
	c := NewNode("c")
	d := NewNode("d")
	a.next = b
	b.next = c
	c.next = d

	fmt.Println(linkedListValues(a))
}
